datafusion/src/optimizer/constant_folding.rs

```diff
@@ -217,6 +218,29 @@ impl<'a> ExprRewriter for ConstantRewriter<'a> {
                     .query_execution_start_time
                     .timestamp_nanos(),
             ))),
+            Expr::ScalarFunction {
+                fun: BuiltinScalarFunction::ToTimestamp,
+                args,
+            } => {
+                if !args.is_empty() {
+                    match &args[0] {
+                        Expr::Literal(ScalarValue::Utf8(Some(val))) => {
+                            Expr::Literal(ScalarValue::TimestampNanosecond(
+                                string_to_timestamp_nanos(val).ok(),
```

Comment:
I don't think the `ok()` here preserves the same semantics.

I tested what happens when an invalid argument is passed to `to_timestamp` on master (via `cargo run -p datafusion-cli`):
```
> select to_timestamp('foo');
ArrowError(ExternalError(Execution("Error parsing \'foo\' as timestamp")))
```

On this branch
```
> select to_timestamp('foo');
+---------------------------+
| TimestampNanosecond(NULL) |
+---------------------------+
|                           |
+---------------------------+
```


I suggest that if the `to_timestamp(..)` call would generate a runtime error, then the optimizer pass should just leave the expression as is (aka don't do the rewrite and leave the original expression)

Reply:
Makes sense. I'll get this fixed today 👍

Reply:
Thank you